OSDI action handler and server for import and people, questions, answers, messages

joshco opened the pull request:
# Fixes 
#597 OSDI record_canvass_helper action handler
#949 API to push contacts into a campaign

## Description

This PR contains an OSDI action handler to submit canvass responses like Question Answers, Tags (aka in VAN terms Survey Question Responses, Activist Codes).  The handler will automatically download the available questions/SQs and tags/ACs and show them in the interaction script editing form as actions.
This can be used to submit data to EveryAction/VAN, Action Network, CiviCRM (in-progress) and other systems that support the OSDI Simple Organizing Profile. ( https://github.com/opensupporter/osdi-docs/pull/337 )

This also contains OSDI Server functionality that lets other OSDI clients read data from Spoke in OSDI format, as well as import contacts with batch mode.
https://github.com/opensupporter/osdi-docs/pull/335

# Checklist:

- [x] I have manually tested my changes on desktop and mobile
- [x] The test suite passes locally with my changes
- [x] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [ ] [My change is 300 lines of code or less](https://github.com/MoveOnOrg/Spoke/blob/main/CONTRIBUTING.md#your-first-code-contribution), or has a documented reason in the description why it’s longer
- [x] I have made any necessary changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] My PR is labeled [WIP] if it is in progress

This pull request is more than 300 lines of code.  This is due to the functionality added, as well as incorporating several rounds of feedback, and customer requests.

-------
@joemcl @lperson  I ended up creating a more expansive OSDI implementation.  In addition to supporting people import and collection browsing, I mapped spoke interaction steps, question_response into OSDI Questions and Answers.  I also expose the messages collection, assignments, and users. as well as the relationship links between them.

In addition to automating import, this can be used to simplify integrations moving Q&A data into other systems like VAN et al.  A script in a nightly cron job could suck down new answers, associated people, and do the mapping between the spoke Q&A and the VAN survey questions, or other logic and push them into VAN et al.

I set up a test server with some sample data and a public API for people to play with here:

Browse to http://spoke.dev.joshco.org/osdi

While looking at the HAL browser, try the following walkthrough

1. Click the link for the organization you want to browse
2. See the list of campaigns
3. Choose the campaign you want to browse
4. You should now be at the API Entry Point for the Campaign
5. Click the link for 'osdi:people' and see the first page of people results
6. Choose a person, and look at it's links.  You can navigate to that person's Answers (question responses), or messages
7. Go back to the AEP. (or Click "Go To Entry Point" in the top navbar to start over)
8. Click the link for 'osdi:answers' and see the recently added answers.  From each answer you can navigate to the person (campaign contact) or the question itself (derived from interaction steps)

> You can also start navigating users -> assignments -> messages

More docs here: https://github.com/joshco/Spoke/blob/osdi_key/docs/OSDI.md

joshco commented:
I created a free script that automates moving recent question_responses into VAN with a customization mapping between Spoke and VAN Q&A.  You can put it in a nightly cron job on your Spoke server to update VAN on the day's work.
https://github.com/opensupporter/queersync

schuyler1d commented on src/server/api/osdi/api-auth.js:
```diff
@@ -0,0 +1,62 @@
+  if (osdi_api_token) {
+    apiKeyInHeader = osdi_api_token;
+  }
+  const bypass=(features.bypass==apiKeyInHeader);
```
This should be `===` and I'm nervous about this bypass feature -- if this is for testing, maybe we can connect it to the `global.TEST_ENVIRONMENT === '1'`

joshco replied:
bypass is gone now too.  triple equality added

schuyler1d commented on src/server/api/osdi/api-auth.js:
```diff
@@ -0,0 +1,62 @@
+
+  if (!(('authorization' in req.headers)
+      || ('osdi-api-token' in req.headers))) {
+    sendUnauthorizedResponse(res)
```
Can we also bail early if apiKey isn't present?

Also, I worry a bit about the difference in response if we enable this feature or not -- maybe we should be returning 404 in these cases?  Thoughts?

joshco replied:
this changed too.  for an auto failure we just return 403 Forbidden.  If OSDI is disabled, it will return 510.  Both include a json body with hints on why.  eg:
__Access Denied - 403 Forbidden__
```
{
  "osdi:error": {
    "response_code": 403,
    "error_description": "You need to use an OSDI-API-Token HTTP header containing your token, or you need to be logged in with sufficient privileges.",
    "auth_status": {
      "token_status": "INVALID_OSDI_TOKEN",
      "user_status": "LOGGED_OUT",
      "authenticated": false,
      "osdi_enabled": true
    }
  }
}
```
__OSDI Disabled - 510 Not Extended__
```
{
  "osdi:error": {
    "response_code": 510,
    "error_description": "OSDI is not enabled for this resource"
  }
}
```

joshco commented:
@schuyler1d  Thanks for the kind remark.  I've reviewed your feedback and agree.  I've also discussed the issue with other OSDI stakeholders, and have a proposal for next steps.

The current pull request includes some functionality for vision's sake.  It also includes the bypass and public API switches which were meant to reduce friction for those who wanted to experience it for themselves using the demo server.  I'll remove those, or refactor it as skyler suggested. 

The current PR also goes beyond the settled resources in OSDI based on my own improvisations of  assignments and messages.  I'll remove these resources and produce an updated, more constrained PR, incorporating Sky and others' feedback.  This way we could have working code this cycle.

In parallel, settling on the definition of what these new resources should be a democratic consensus among OSDI stakeholders, with leadership from those who build products that implement these features.  

Augustus Franklin @augfrank , from CallHub (which also implements similar features) has offered to serve as a convener of an OSDI a breakout group on these resources. It would be great if someone from Spoke would participate as well as Hustle, Relay and any other implementers.

augfrank commented:
@joshco @schuyler1d Yes, I'd love to help. It would be good to setup a common way to represent messages and its related resources.

ibrand commented:
@joshco @schuyler1d @augfrank What's the status on this work?

joshco commented on src/server/api/osdi/contactsApi.js:
```diff
@@ -0,0 +1,256 @@
+
+      resp = {contacts: {deleted}}
+    } else if (req.method === 'POST') {
+      const osdi_body = req.body;
```
Both of these already have it. See line 56  - that's the other short-circuit :)
https://github.com/MoveOnOrg/Spoke/blob/69b2ed851326f0406f2167ae9f022478c5eac332/src/server/api/osdi/contactsApi.js#L56

src/server/api/osdi/contactsApi.js#L55

joshco commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "hal"]
```
I dont think this is a dev dependency.  The HAL browser is essentially the API console for humans to experiment, learn, and guide them in their scripting.
Ideally I'd like to avoid hard coding it into Spoke, as I make periodic updates to it in my github fork.
What alternatives can you think of?

joshco replied:
I could clone it as part of the yarn install.  what do you think about that?

joshco replied:
sounds great to me.  
https://github.com/opensupporter/hal-browser.git

joshco commented on src/server/api/osdi/api-auth.js:
```diff
@@ -0,0 +1,62 @@
+}
+
+function publicApi(features) {
+  if (features.publicApi === true || process.env.PUBLIC_API === 'true') {
```
I've taken this out in the new version.  Now the authentication will allow a logged in user to browse the API with HAL browser, without futzing with an API key.  Right now the role required is ADMIN.  My thought is that the owner might want to give tech supervols access to do some scripting.

joshco commented on src/server/api/osdi/contactsApi.js:
```diff
@@ -0,0 +1,256 @@
+        }
+      }
+    } else if (req.method === 'DELETE') {
+      const deleted = await r
```
see next...

joshco commented:
@ibrand This is almost complete.  I've made most of the changes Skyler requested.  I'll push a branch today.

joshco commented on src/server/api/osdi/contactsApi.js:
```diff
@@ -0,0 +1,256 @@
+          }
+      )
+
+      const contactsToSave = await Promise.all(contactsToSavePromises)
```
looking into that

joshco commented on src/server/index.js:
```diff
@@ -173,6 +178,9 @@ app.get('/graphiql', graphiqlExpress({
 }))
 
+
+osdiInitialize(app)
```
I put a toggle to turn OSDI off and on, like enforce texter hours.  
![image](https://user-images.githubusercontent.com/290314/63641944-ef0b7900-c684-11e9-9e1f-b07633d4c302.png)

and
![image](https://user-images.githubusercontent.com/290314/63641950-00548580-c685-11e9-9f4c-b94afc20b144.png)

and

![image](https://user-images.githubusercontent.com/290314/63641954-0d717480-c685-11e9-8ba2-50e7cb6d6ee8.png)

schuyler1d commented:
I haven't reviewed but the last time I reviewed this these were my blockers:
* the static assets need to be built/setup however other static assets are with webpack, etc. -- this is the github 'clone' project -- @ibrand -- you might have thoughts opinions on how best to link this.  Unfortunately, it's been a while since anyone has touched static asset stuff, but e.g. AWS Lambda will not serve static assets, and so without doing this 'correctly' it won't work.  It also is for best practice -- separating serving static assets from server responses.
* Without a default-off environment variable, no endpoints or code should be loaded -- i.e. this should be tested in src/server/index.js before passing `app` into the code.
* My first review identified several security vulnerabilities -- I think these were closed since then, but I'd appreciate someone else reviewing the code and 1. confirming the identified ones are closed, 2. someone else do a review and affirm that.

Sorry, I haven't had time to review this (or do much else regarding Spoke) recently -- have a project launch at work I'm battling to keep 'on time'.

But ++ to @joshco 's work -- this could open up some great ways of linking into Spoke from other systems.

schuyler1d commented:
updated my list above, after remembering another:
I'd like to see at least some minimal testing -- maybe test of one or two endpoints -- that way if people create more tests they have the pattern they need to do so. @ibrand -- defer to you on priority for this.

joshco commented:
Latest update. @schuyler1d @ibrand  I've added the global ENV var, improved the settings dialog.  I've also removed the git submodule for the osdi browser, and instead snapshotted the browser code and am using webpack copy plugin. This will put it in the webpack, and also drop it in the assets directory which will be picked up by the S3 upload script.

https://github.com/joshco/Spoke/blob/osdi_key/webpack/config.js#L21

joemcl commented:
thanks @joshco also worth looking at https://github.com/move-coop/parsons_public in relation to this.  cc: @schuyler1d @ibrand @shakalee14

joshco commented:
@joemcl  Thanks for passing this along; I doubt my response will surprise you. Approaches like this that encourage the use of proprietary APIs disproportionately harm the most marginalized communities in the progressive movement. The result is vendor lock-in where vendors will win sales based on avoiding the higher cost of integration with other products, rather than the merits of the product itself.  This helps the big get bigger due to network effect.  Larger customer organizations, or larger campaigns (like a presidential or $12m statewide gay marriage campaign) can overcome these costs with money or volunteer brute-force. However, smaller and more marginalized campaigns and organisations will not be able to do this as easily, and will stick with the larger "suite" type vendors.  This further disadvantages products from marginalized innovators, because the additional cost of integrating those products will dis-incentivize customers from choosing one of these products based on merit and solidarity.

Given the recent consolidation in our marketplace, IMHO this will make our problems worse not better.

joshco commented:
I've updated the branch with 2 OSDI test cases.
1. get people with apitoken succeeds, and without fails
2. try to get people with osdi disabled and it fails

One weird thing I'm seeing in the tests is that intermittently, I'm getting an error in test_helpers.  It seems to happen when I run the tests for the first time after time has passed.  But then it works.
```
TypeError: Cannot read property 'data' of undefined
at: const organizationId = organization.data.createOrganization.id 
```
@schuyler1d Any ideas why?  Who's the most familiar with the tests?

joshco commented:
@schuyler1d @ibrand @joemcl I've added outbound OSDI (like action handlers) for interaction steps to submit survey question answers or apply tags/activist codes.  By default, it works with VAN / EveryAction unless a different AEP is configured.

I'm looking for dogfooders who would like to give it a spin.

From the updated documentation:

## Outbound OSDI Actions

### Configuration
Outbound OSDI is configured with these environment variables
* __OSDI_OUTBOUND_API_KEY__ Your OSDI API TOKEN
* __OSDI_OUTBOUND_AEP__ [Optional] The URL of your OSDI API Entry Point

By default, if you configured an API TOKEN but no AEP, Spoke will push to VAN / EveryAction.

### Usage
Once configured, when you are editing your interaction steps, Spoke will download the available survey questions and tags (aka Activist Codes in VAN)

In the drop down for actions, you'll see options to apply tags/activist codes or submit responses to survey questions.

#### Dropdown of Choices
![osdi_outbound_choices](https://user-images.githubusercontent.com/290314/66954553-a5a92b80-f02e-11e9-9174-6188d9a6f08a.png)


#### Choice Chosen
![osdi_outbound_choice](https://user-images.githubusercontent.com/290314/66954784-10f2fd80-f02f-11e9-80dc-363833960c3b.png)


### Matching Logic
By default, when pushing via OSDI, Spoke will invoke the `osdi:person_signup_helper` endpoint on the OSDI server to match or create a corresponding OSDI person.  It will do based on fields in spoke for mobile phone number, zip, name etc.

Once matched, Spoke will invoke `osdi:record_canvass_helper` to apply tags/activist codes and submit responses to survey questions. 

Spoke will also save the remote OSDI Identifier in a contact custom field named `osdi_identifier` to avoid matching on subsequent actions for teh same contact.

To improve matching, when you import into spoke, include `email` as a custom field.  

If you are importing contacts that you got from the OSDI system itself (eg VAN), include the OSDI Identifier itself (eg VANID) in a custom field named `osdi_identifier`

If `osdi_identifier` is present, then Spoke will skip the `osdi:person_signup_helper` and just to the `osdi:record_canvass_helper`

joemcl commented:
hey @joshco @schuyler1d @ibrand and adding in @lperson - sorry I am just now jumping in on this again. Can y'all possibly collaborate and fast track getting this merged? It's great stuff, and I'd *really* love to use it for an effort that I'm working on - it's for an election happening this November.

joshco commented:
@joemcl I've merged in main, as of today.

joemcl commented:
@joshco thanks so much and I'll test that out today/tonight on a Spoke instance.

joemcl commented:
So I deployed this tonight on to a production site and will test in the a.m.., after finishing set up. Just a note that the doc doesn't say anything about `OSDI_MASTER_ENABLE` env var, however I added it and set it to `true` which works, so now I see the `OSDI API Access` toggle....

joshco commented:
@joemcl  I did some testing with the other CRM and made an update which pertains to you.  You'll want to pull this.
I've added an optional environment variable OSDI_SERVER_CONTENT_TYPE which sets the content type of responses.  The default is application/json, but some CRMs need application/hal+json.  So you should set that to application/hal+json in your environment.  
I've updated the documentation to include both variables.

joemcl commented:
thanks @joshco can't deploy that latest update til late tonight probably. But I can attempt to set up a VAN connection with what I have now, correct?

joshco commented:
@joemcl yes, that's correct. You can set the outbound push to VAN.

joshco commented:
@joemcl added odata date filtering, so queersync can be more efficient.  I've also updated the civi extension to improve the AEP and PSH endpoint.  

https://lab.civicrm.org/joshalot/CiviCRM-OSDI/tree/webhook_updates

Updated queersync to run as signup only (when survey questions aren't needed in other systems)

You should now be able to have it autopush to VAN via interaction steps/scripts, and sync nightly with Civi via queersync.

joshco commented:
Improvements based on customer testing and feedback to make it easier to use and set up.  Thanks to @joemcl 
* Added `yarn osdi-info` script to check config, view available information from remote OSDI Server
* Added additional environment variables to option in logging of OSDI wire level traces
* UI improvements 

@lperson Can we update the google doc script importer so the customer can put a token for an `answer_action` in the google doc, and your script will stick it in the DB `answer_action` column?  
Customers can now use the osdi-info CLI to discover available actions (activist codes, survey questions, tags) on their remote OSDI server.  It will give them a token they can use in the google doc.

@ibrand @augfrank @schuyler1d

schuyler1d commented:
contact-loaders sadly make this structure obsolete.
